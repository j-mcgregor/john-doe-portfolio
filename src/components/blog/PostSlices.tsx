import React from 'react'
import {
    PRISMIC_PRIMARY_Image_with_Caption,
    PRISMIC_PRIMARY_Quote,
    PRISMIC_PRIMARY_Text,
} from '../../types/interfaces/blog'
import { ImageCaption, Quote, Text } from '../../components/shared/slices'
import createKey from '../../utils/createKey'

export interface PostSlicesProps {
    // INTERSECTION TYPES
    // https://www.typescriptlang.org/docs/handbook/advanced-types.html#intersection-types
    slices: Array<
        PRISMIC_PRIMARY_Image_with_Caption &
            PRISMIC_PRIMARY_Quote &
            PRISMIC_PRIMARY_Text
    >
}

// Sort and display the different slice options
const PostSlices: React.FC<PostSlicesProps> = ({ slices }) => {
    const postSlices = slices.map((slice, index) => {
        switch (slice.type) {
            case 'text':
                return (
                    <div
                        key={createKey(slice.type, index)}
                        className="homepage-slice-wrapper"
                    >
                        <Text slice={slice} />
                    </div>
                )

            case 'quote':
                return (
                    <div
                        key={createKey(slice.type, index)}
                        className="homepage-slice-wrapper"
                    >
                        <Quote slice={slice} />
                    </div>
                )

            case 'image_with_caption':
                return (
                    <div
                        key={createKey(slice.type, index)}
                        className="homepage-slice-wrapper"
                    >
                        <ImageCaption slice={slice} />
                    </div>
                )

            default:
                return <div key={createKey(slice.type, index)} />
        }
    })

    return <div>{postSlices}</div>
}

export default PostSlices
