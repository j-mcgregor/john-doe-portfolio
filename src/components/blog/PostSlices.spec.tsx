import React from 'react'
import { render } from '@testing-library/react'
import PostSlices, { PostSlicesProps } from './PostSlices'
import {
    textSlice,
    imageSlice,
    quoteSlice,
} from '../../__mocks__/data/blogPage'

describe('PostSlices', () => {
    it('should render a Text Slice', () => {
        const props = {
            slices: [textSlice('Lorem ipsum dolor sit amet')],
        } as PostSlicesProps

        const { baseElement, getAllByText } = render(<PostSlices {...props} />)

        expect(baseElement).toMatchSnapshot()
        expect(getAllByText('Lorem ipsum dolor sit amet')).toBeDefined()
    })

    it('should render an Image Slice', () => {
        const props = {
            slices: [imageSlice(0)],
        } as PostSlicesProps

        const { baseElement } = render(<PostSlices {...props} />)

        expect(baseElement).toMatchSnapshot()
        expect(baseElement.getElementsByTagName('img')[0].src).toBe(
            'http://fillmurray.com/300/300'
        )
    })

    it('should render a Quote Slice', () => {
        const props = {
            slices: [quoteSlice('And I quoteth thus...')],
        } as PostSlicesProps

        const { baseElement, getAllByText } = render(<PostSlices {...props} />)

        expect(baseElement).toMatchSnapshot()
        expect(getAllByText('And I quoteth thus...')).toBeDefined()
    })

    it('should render all slices', () => {
        const props = {
            slices: [
                textSlice('Lorem ipsum dolor sit amet'),
                imageSlice(0),
                quoteSlice('And I quoteth thus...'),
            ],
        } as PostSlicesProps

        const { baseElement, getAllByText } = render(<PostSlices {...props} />)

        expect(baseElement).toMatchSnapshot()
        expect(getAllByText('Lorem ipsum dolor sit amet')).toBeDefined()
        expect(getAllByText('And I quoteth thus...')).toBeDefined()
        expect(baseElement.getElementsByTagName('img')[0].src).toBe(
            'http://fillmurray.com/300/300'
        )
    })

    it('should render empty div if type !== match', () => {
        const props = {
            slices: [
                {
                    type: 'batman',
                },
            ],
        } as PostSlicesProps

        const { baseElement } = render(<PostSlices {...props} />)

        expect(baseElement).toMatchSnapshot()
    })
})
