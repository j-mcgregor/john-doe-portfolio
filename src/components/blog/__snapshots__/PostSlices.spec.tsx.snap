// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostSlices should render a Quote Slice 1`] = `
<body>
  <div>
    <div>
      <div
        class="homepage-slice-wrapper"
      >
        <div
          class="post-quote container"
        >
          <blockquote>
            <p>
              And I quoteth thus...
            </p>
          </blockquote>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`PostSlices should render a Text Slice 1`] = `
<body>
  <div>
    <div>
      <div
        class="homepage-slice-wrapper"
      >
        <div
          class="post-text container"
        >
          <div>
            <p>
              Lorem ipsum dolor sit amet
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`PostSlices should render all slices 1`] = `
<body>
  <div>
    <div>
      <div
        class="homepage-slice-wrapper"
      >
        <div
          class="post-text container"
        >
          <div>
            <p>
              Lorem ipsum dolor sit amet
            </p>
          </div>
        </div>
      </div>
      <div
        class="homepage-slice-wrapper"
      >
        <div
          class="post-image container"
        >
          <figcaption
            class="block-img"
          >
            <img
              alt="alt-0"
              src="http://fillmurray.com/300/300"
            />
            <figcaption
              class="image-label"
            >
              caption-0
            </figcaption>
          </figcaption>
        </div>
      </div>
      <div
        class="homepage-slice-wrapper"
      >
        <div
          class="post-quote container"
        >
          <blockquote>
            <p>
              And I quoteth thus...
            </p>
          </blockquote>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`PostSlices should render an Image Slice 1`] = `
<body>
  <div>
    <div>
      <div
        class="homepage-slice-wrapper"
      >
        <div
          class="post-image container"
        >
          <figcaption
            class="block-img"
          >
            <img
              alt="alt-0"
              src="http://fillmurray.com/300/300"
            />
            <figcaption
              class="image-label"
            >
              caption-0
            </figcaption>
          </figcaption>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`PostSlices should render empty div if type !== match 1`] = `
<body>
  <div>
    <div>
      <div />
    </div>
  </div>
</body>
`;
